src/font_file_lister_gdi.cpp

```diff
@@ -210,90 +175,112 @@ CollectionResult GdiFontFileLister::GetFontPaths(std::string const& facename, in
-				continue;
-			if (indices[i] == SHRT_MAX)
-				ret.missing += utf16characters[i];
+	for (int character : characters) {
+		if (!font_face_3_sh->HasCharacter((UINT32)character)) {
+			ret.missing += character;
 		}
 	}
-	else if (hr == S_FALSE) {
-		for (size_t i = 0; i < utf16characters.size(); ++i) {
-			// Uniscribe doesn't report glyph indexes for non-BMP characters,
-			// so we have to call ScriptGetCMap on each individual pair to
-			// determine if it's the missing one
-			if (U16_IS_LEAD(utf16characters[i])) {
-				hr = ScriptGetCMap(dc, &cache, &utf16characters[i], 2, 0, &indices[i]);
-				if (hr == S_FALSE) {
-					ret.missing += utf16characters[i];
-					ret.missing += utf16characters[i + 1];
-				}
-				++i;
-			}
-			else if (indices[i] == 0) {
-				ret.missing += utf16characters[i];
-			}
+#else
+	IDWriteFont* font;
+	if (FAILED(font_collection_sh->GetFontFromFontFace(font_face_sh, &font)))
+	{
+		return ret;
+	}
+	agi::scoped_holder<IDWriteFont*> font_sh(font, [](IDWriteFont* p) { p->Release(); });
+
+	BOOL exists;
+	HRESULT hr;
+	for (int character : characters) {
+		hr = font_sh->HasCharacter((UINT32)character, &exists);
+		if (FAILED(hr) || !exists) {
+			ret.missing += character;
 		}
 	}
-	ScriptFreeCache(&cache);
+#endif
+
+	UINT32 file_count = 1;
+	IDWriteFontFile* font_file;
+	// DirectWrite only supports one file per face
+	if (FAILED(font_face_sh->GetFiles(&file_count, &font_file)))
+	{
+		return ret;
+	}
+	agi::scoped_holder<IDWriteFontFile*> font_file_sh(font_file, [](IDWriteFontFile* p) { p->Release(); });
+
+	IDWriteFontFileLoader* loader;
+	if (FAILED(font_file_sh->GetLoader(&loader)))
+	{
+		return ret;
+	}
+	agi::scoped_holder<IDWriteFontFileLoader*> loader_sh(loader, [](IDWriteFontFileLoader* p) { p->Release(); });
+
+	IDWriteLocalFontFileLoader* local_loader;
+	if (FAILED(loader_sh->QueryInterface(__uuidof(IDWriteLocalFontFileLoader), (void**)&local_loader)))
+	{
+		return ret;
+	}
+	agi::scoped_holder<IDWriteLocalFontFileLoader*> local_loader_sh(local_loader, [](IDWriteLocalFontFileLoader* p) { p->Release(); });
+
+	LPCVOID font_file_reference_key;
+	UINT32 font_file_reference_key_size;
+	if (FAILED(font_file_sh->GetReferenceKey(&font_file_reference_key, &font_file_reference_key_size)))
+	{
+		return ret;
+	}
+
+	UINT32 path_length;
+	if (FAILED(local_loader_sh->GetFilePathLengthFromKey(font_file_reference_key, font_file_reference_key_size, &path_length)))
+	{
+		return ret;
+	}
+
+	WCHAR* path = new WCHAR[path_length + 1];
+	if (path == nullptr) {
+		return ret;
+	}
+	agi::scoped_holder<WCHAR**> path_sh(&path, [](WCHAR** p) { delete[] *p; });
+
+	if (FAILED(local_loader_sh->GetFilePathFromKey(font_file_reference_key, font_file_reference_key_size, path, path_length + 1)))
+	{
+		return ret;
+	}
+
+	// DirectWrite always return the filepath in upper case. Ex: "C:\WINDOWS\FONTS\ARIAL.TTF"
+	DWORD normalized_path_length;
+	if (!normalizeFilePathCase(&path, &normalized_path_length)) {
```

Comment:
It'd be more consistent to use `path_sh` rather than `&path` here (and similarly in the `ag::fs::path` call below).

Reply:
Addressed in commit 38a5a99af420aedf926cb6b08cc54c24c8e6d2ad
Now, I don't use WCHAR anymore, so it make this remark obsolete.